In [3]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_curve
import math

%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [7]:
data = pd.read_csv("CS570/data/breast_cancer.csv")

In [8]:
data.head()

Unnamed: 0  id number  clump_thickness  uniformity_of_cell_size  \
0           0    1000025                5                        1   
1           1    1002945                5                        4   
2           2    1015425                3                        1   
3           3    1016277                6                        8   
4           4    1017023                4                        1   

   uniformity_of_cell_shape  marginal_adhesion  epithelial_cell_size  \
0                         1                  1                     2   
1                         4                  5                     7   
2                         1                  1                     2   
3                         8                  1                     3   
4                         1                  3                     2   

   bare_nuclei  bland_chromatin  normal_nucleoli  mitoses  malignant  
0            1                3                1        1          0  
1           10                3                2        1          0  
2            2                3                1        1          0  
3            4                3                7        1          0  
4            1                3                1        1          0

In [9]:
data = data.drop(['Unnamed: 0', 'id number'], axis=1)

In [10]:
y = data.pop("malignant")

In [11]:
X_train, X_test, y_train, y_test = train_test_split(data, y, test_size=.2, random_state=42)

In [12]:
n_estimators = [300,400,500]
max_features = ['auto', 'sqrt','log2']
min_samples_split = [3,5,7]


rfc = RandomForestClassifier(n_jobs=1)
estimator = GridSearchCV(rfc,
                         dict(n_estimators=n_estimators,
                              max_features=max_features,
                              min_samples_split=min_samples_split
                              ), cv=None, n_jobs=-1)

In [14]:
estimator.fit(X_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params={}, iid=True, n_jobs=-1,
       param_grid={'min_samples_split': [3, 5, 7], 'max_features': ['auto', 'sqrt', 'log2'], 'n_estimators': [300, 400, 500]},
       pre_dispatch='2*n_jobs', refit=True, scoring=None, verbose=0)

In [20]:
estimator.best_estimator_

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=7,
            min_weight_fraction_leaf=0.0, n_estimators=500, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [21]:
best_rfc = estimator.best_estimator_

### Accuracy

In [22]:
accuracy = accuracy_score(y_test, best_rfc.predict(X_test))
print ("Accuracy: ", accuracy)

('Accuracy: ', 0.9642857142857143)


### Precision and Recall

In [23]:
print (classification_report(y_test, best_rfc.predict(X_test)))

             precision    recall  f1-score   support

          0       0.97      0.98      0.97        95
          1       0.95      0.93      0.94        45

avg / total       0.96      0.96      0.96       140



### AUC

In [24]:
roc = roc_auc_score(y_test, best_rfc.predict_proba(X_test)[:,1])
print ("AUC Score: ", roc)

('AUC Score: ', 0.99578947368421056)


### K-Fold Cross Validation

In [25]:
from sklearn import cross_validation
scores = cross_validation.cross_val_score(best_rfc, data, y, cv=10)

In [26]:
mean_score = scores.mean()
std_dev = scores.std()
std_error = scores.std() / math.sqrt(scores.shape[0])
ci =  2.262 * std_error
lower_bound = mean_score - ci
upper_bound = mean_score + ci

print ("Score is %f +/-  %f" % (mean_score, ci))
print ('95 percent probability that if this experiment were repeated over and over the average score would be between %f and %f' % (lower_bound, upper_bound))

Score is 0.970059 +/-  0.017827
95 percent probability that if this experiment were repeated over and over the average score would be between 0.952233 and 0.987886


### Model's Performance
In short, this model's performance is pretty darn good. Working top down, an accuracy, while a basic measure of performance, of 96.42% is good. Precision of 0.96 and a recall of 0.96 similarly supports accuracy conclusion that this model is good. The AUC tells the same tale. An AUC score of 0.9957 is very strong. 